src/twisted/web/test/test_websocket.py

```diff
@@ -0,0 +1,344 @@
+
+        def bytesMessageReceived(self, data: bytes) -> None:
+            if data == b"request":
+                self.transport.sendBytesMessage(b"\x00resp\x01onse\xff")
+            else:
+                self.bDeferred.callback(data)
+
+        def textMessageReceived(self, data: str) -> None:
+            if data == "request":
+                self.transport.sendTextMessage("response")
+            else:
+                self.deferred.callback(data)
+
+        def pongReceived(self, payload: bytes) -> None:
+            self.pongs.append(payload)
+
+        def sendRequest(self) -> Deferred[str]:
+            """
+            Send a text message to the server and expect a response.
+            """
+            self.deferred: Deferred[str] = Deferred()
+            self.transport.sendTextMessage("request")
+            return self.deferred
+
+        def bytesRequest(self) -> Deferred[bytes]:
+            """
+            Send a bytes message to the server and expect a response.
+            """
+            self.bDeferred: Deferred[bytes] = Deferred()
+            self.transport.sendBytesMessage(b"request")
+            return self.bDeferred
+
+        def goodbye(self) -> None:
+            self.transport.loseConnection()
+
+    class MyFactory(WebSocketServerFactory[MyWSP]):
+        fixture: WebSocketFixture[Any]
+
+        def buildProtocol(self, request: IRequest) -> MyWSP:
+            new = MyWSP()
+            self.fixture.servers.append(new)
+            return new
+
+    class MyClientFactory(WebSocketClientFactory[MyWSP]):
+        def buildProtocol(self, uri: str) -> MyWSP:
+            return MyWSP()
+
+
+@dataclass
+class WebSocketFixture(Generic[WSP]):
+    clientFactory: WebSocketClientFactory[WSP] = field()
+    reactor: MemoryReactorClock = field(default_factory=MemoryReactorClock)
+    resource: Resource = field(default_factory=Resource)
+    portNumber: int = 80
+    servers: list[WSP] = field(default_factory=list)
+    slowResource: SlowResource = field(default_factory=SlowResource)
+
+    @classmethod
+    def new(cls, clientFactory: WebSocketClientFactory[WSP]) -> WebSocketFixture[WSP]:
+        self = cls(clientFactory)
+        serverFactory = MyFactory()
+        serverFactory.fixture = self
+        self.resource.putChild(b"connect", WebSocketResource(serverFactory))
+        self.resource.putChild(
+            b"resource", Data(b"some-data", "application/octet-stream")
+        )
+        self.resource.putChild(b"processing", WeirdResource())
+        self.resource.putChild(b"slow", self.slowResource)
+        self.reactor.listenTCP(
+            self.portNumber, Site(self.resource, reactor=self.reactor)
+        )
+        return self
+
+    async def connect(self, uri: str = "http://localhost:80/connect") -> WSP:
+        client = WebSocketClientEndpoint.new(self.reactor, uri)
+        return await client.connect(self.clientFactory)
+
+    def complete(self, greet: bool = True) -> IOPump:
```

Comment:
I find the `complete` name very generic..and there is no documentation for the `greet` argument.

What does `complete` means ?

I guess that all data is exchange between client and server ... but are we also closing the connection?

Maybe something like this... to hint that this is iosim API ?

```suggestion
    def finalPump(self, greet: bool = True) -> IOPump:
```

Reply:
I am not sure what "final" would mean here, but I expanded the docstring.